feat: Add return and exception event tracing

AnshKumarTripathi opened the pull request:
##  Summary

This PR implements Issue #3: Trace return and exception events, significantly enhancing the debugging capabilities of spewer.

##  What's Changed

### **New Features:**
-  **Return Event Tracing**: Complete function execution flow (entry + return)
-  **Exception Event Tracing**: Full exception visibility with details
-  **Configuration Options**: `trace_returns` and `trace_exceptions` parameters
-  **Both Tracing Modes**: Works with `functions_only=True` and line-by-line tracing
-  **Nested Function Support**: Complete call stack tracing

### **API Changes:**
- Added `trace_returns: bool = True` parameter to `spew()` and `SpewContext`
- Added `trace_exceptions: bool = True` parameter to `spew()` and `SpewContext`
- Enhanced `SpewConfig` with new boolean options and validation

### **Implementation Details:**
- Enhanced `TraceHook.__call__()` to handle return and exception events
- Added `_handle_function_return()` and `_handle_function_exception()` methods
- Added `_handle_line_return()` and `_handle_line_exception()` methods
- Updated configuration validation in `SpewConfig.__post_init__()`

## 🧪 Testing

### **Test Script:**
```python
# Run the comprehensive test
python test_issue_3_implementation.py

AnshKumarTripathi committed:
feat: Add return and exception event tracing

- Add trace_returns and trace_exceptions configuration options
- Implement return event handling for complete function execution flow
- Implement exception event handling for better debugging visibility
- Add _handle_function_return() and _handle_function_exception() methods
- Add _handle_line_return() and _handle_line_exception() methods
- Update SpewConfig with new boolean options and validation
- Update spew() function and SpewContext with new parameters
- Support both functions_only and line-by-line tracing modes
- Add comprehensive test script demonstrating new functionality

Fixes #3: Trace return and exception events

AnshKumarTripathi committed:
fix: Address review feedback - add unit tests, update README, setup pre-commit

- Add comprehensive unit tests for return and exception event handling
- Remove system test file as requested by maintainer
- Update README with new features and examples
- Set up pre-commit hooks for code quality
- Fix all linting and formatting issues
- Increase test coverage from 23 to 51 tests

Addresses maintainer feedback for PR #8
Fixes #3: Trace return and exception events

AnshKumarTripathi committed:
fix: Address ruff linting errors and improve test coverage

- Fix unused variable warnings in exception handling (exc_tb -> _)
- Add comprehensive test coverage for missing lines in trace.py
- Cover linecache.getline() and inspect.getsourcelines() code paths
- Add 5 new test cases to cover all missing coverage lines
- Improve overall test coverage to meet project standards
- All 56 tests passing with 0 linting errors

Addresses maintainer feedback for PR #8

AnshKumarTripathi committed:
fix: Reorganize tests into test_spewer.py as requested - Removed test_trace_new_features.py and test_coverage_gaps.py - Added 20 unit tests to TestTraceHook class - Added 7 integration tests to TestIntegration class - All 53 tests passing (increased from 23) - Organized tests into existing 4 test classes as maintainer requested

AnshKumarTripathi committed:
fix: Change default values for trace_returns and trace_exceptions to False - Changed trace_returns default from True to False - Changed trace_exceptions default from False to False - Removed 'Configurable tracing' line from README features - Updated all documentation to reflect new defaults - Makes new features opt-in rather than opt-out for better backward compatibility

AnshKumarTripathi committed:
fix: Update spew() function defaults to False

- Changed trace_returns default from True to False in spew() function
- Changed trace_exceptions default from True to False in spew() function
- Now consistent with SpewContext and SpewConfig defaults
- Completes the change to make new features opt-in across entire codebase

AnshKumarTripathi commented:
@Agent-Hellboy I added the changes requested in the new commit.
Also i had some question:
How does Spewer handle memory usage when tracing long-running applications? Are there any built-in safeguards against memory leaks?

Agent-Hellboy commented:
Hi @AnshKumarTripathi , please add unittest, you have added intergation test. please 

> Also can you tell me how to run this all bot tests that happens in GitHub? So i can run it locally first and push only the changes that are fully passed. So it saves me some embracement :)

got to project root and do pre-commit install after installing https://pre-commit.com/
when ever you run `git commit` it will run the commit hook. 


check tox.ini and ci files to understand how does ci run here, please remove AI genarted test script keep it in test_spewer only.

Agent-Hellboy commented:
> How does Spewer handle memory usage when tracing long-running applications? Are there any built-in safeguards against memory leaks?

you can use `memory_profiler` library to check this. I don't see an fd leak here, but please check it though the library i recommended.

Agent-Hellboy commented:
there are 4 classes inside the test file please check in which class your test should go.

AnshKumarTripathi commented:
@Agent-Hellboy Now when I used pre-commit, then i found out it was just a linter error and was not a super duper blocker.
Also thanks for not getting mad

Agent-Hellboy commented:
Hi @AnshKumarTripathi please test your changes against a script with external dependencies. 

If you can please check codecov report and see if you can improve it or not. Please don’t use AI to write unit test.

Agent-Hellboy commented:
> Hi @Agent-Hellboy Sorry for going for two weeks without telling, but i met with an accident so was not able to do it.
> I hope this issue is not closed, as I am still working on this, and probably I will complete this by today or tomorrow.

No problem at all, please take care. Your contribution will be counted; someone will add a patchset, or I will do it myself.

AnshKumarTripathi committed:
fix: Add comprehensive test coverage for return/exception event tracing

- Add 7 new tests covering line-mode return/exception branches
- Add tests for .pyc/.pyo file handling edge cases
- Add integration tests with real Python code execution
- Fix incorrect default values in README documentation
- Coverage improved from 92% to 93% (60 tests, was 53)

Addresses codecov failures in PR #8

Agent-Hellboy commented:
> I used the existing test infrastructure and real Python code execution, not AI-generated boilerplate.

Okay, add a screenshot using this library.

AnshKumarTripathi commented:
> > I used the existing test infrastructure and real Python code execution, not AI-generated boilerplate.
> 
> Okay, add a screenshot using this library.

<img width="1135" height="715" alt="demo_tracing" src="https://github.com/user-attachments/assets/87249267-b704-422b-9b0b-565a28cf5a95" />

Did you mean this? I tired

The screenshot shows:
- **Return events** being traced with values (`calculate_sum() -> 30`)
- **Exception events** being traced (`divide_numbers() -> ValueError(...)`)
- **Both features** working together seamlessly

AnshKumarTripathi commented:
Also I just looked that my PR has some conflicts, so let me know what changes i need to do.

AnshKumarTripathi committed:
Merge branch 'main' into feature/issue-3-return-exception-events
